File: src/main/java/core/procedures/math/Sqrt.java
package core.procedures.math;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMBigComplex;

public final class Sqrt extends AFn {

  public Sqrt() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Number.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "sqrt";
  }

  @Override
  public Number apply1(Object arg) {
    return sqrt((Number) arg);
  }

  public static Number sqrt(Number number) {
    if (number instanceof SCMBigComplex) {
      if (((SCMBigComplex)number).getIm().signum()== 0) {
        return sqrt(((SCMBigComplex)number).getRe());
      }
      return ((SCMBigComplex)number).sqrt();
    }
    return Math.sqrt(number.doubleValue());
  }
}

File: src/main/java/core/procedures/math/Sqrt.kt
package core.procedures.math

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Complex
import core.utils.Utils
import java.math.BigDecimal
import kotlin.math.sqrt

class Sqrt : AFn<Number?, Number>(name = "sqrt", isPure = true, arity = Exactly(1),
                                  mandatoryArgsTypes = arrayOf(Number::class.java)) {

    override operator fun invoke(arg: Number?): Number = when (arg) {
        is Complex -> when (arg.im.signum()) {
            0    -> invoke(arg.re)
            else -> arg.sqrt()
        }
        is BigDecimal -> arg.sqrt(Utils.DEFAULT_CONTEXT)
        else -> sqrt(arg!!.toDouble())
    }
}
